### PR TITLE
Issue-6276: Update iOS sdk

### LINKED ISSUE #5505: Upgrade to iOS 15
**Task (REQUIRED):**
Upgrade Defold to support iOS 15 when it is released.

**Expected outcome (REQUIRED):**
Defold should support iOS 15

### LINKED ISSUE #6276: Update SDK to iOS 15, macOS 12 and XCode 13
It's time for an update.

### PR DESCRIPTION
**Pull request by JCash**
Updated to:
* XCode 13.2.1
* iPhoneOS 15.2
* macOS 12.1

### CODE REVIEW

**Review by britzl — DISMISSED** [review]

**Comment by JCash** [comment]
Also related:
https://github.com/defold/extender/pull/219

File: build_tools/sdk.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env python
```
Comment (JCash) [inline build_tools/sdk.py:1]: Trying to make it easier to keep the version numbers in a single file.
Also, the goal is to allow for finding tools on the local host as a fallback.
(I.e. not requiring to package the tools)

File: build_tools/sdk.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    if platform in ('x86_64-darwin',):
+        return 'macosx'
+    if platform in ('armv7-darwin','arm64-darwin'):
+        return 'iphoneos'
+    if platform in ('x86_64-ios',):
+        return 'iphonesimulator'
+    return 'unknown'
+
+def get_local_darwin_toolchain_path():
+    default_path = '/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain'
```
Comment (JCash) [inline build_tools/sdk.py:74]: I didn't find a good way of locating this folder via script/command line.

File: build_tools/sdk.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+def _get_local_sdk_info(platform):
+    info = {}
+    if platform in ('x86_64-darwin','x86_64-ios','armv7-darwin','arm64-darwin'):
+        info['xcode'] = {}
+        info['xcode']['version'] = get_local_darwin_toolchain_version()
+        info['xcode']['path'] = get_local_darwin_toolchain_path()
+        info[platform] = {}
+        info[platform]['version'] = get_local_darwin_sdk_version(platform)
+        info[platform]['path'] = get_local_darwin_sdk_path(platform)
```
Comment (JCash) [inline build_tools/sdk.py:162]: Each platform has its own requirements, so we allow it to output whatever is needed by the build.py/waf_dynamo.py

File: build_tools/sdk.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    return info
+
+# It's only cached for the duration of one build
+cached_platforms = defaultdict(defaultdict)
+
+def get_sdk_info(sdkfolder, platform):
+    if platform in cached_platforms:
+        return cached_platforms[platform]
+
+    if check_defold_sdk(sdkfolder, platform):
```
Comment (JCash) [inline build_tools/sdk.py:173]: Prioritize the installed packages first...

File: build_tools/sdk.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        result = _get_defold_sdk_info(sdkfolder, platform)
+        cached_platforms[platform] = result
+        return result
+
+    if check_local_sdk(platform):
```
Comment (JCash) [inline build_tools/sdk.py:178]: ...then check for locally installed tools on the host

File: build_tools/waf_dynamo.py [hunk 6]
```diff
@@ -64,33 +65,22 @@ def transform_runnable_path(platform, path):
...
-SDK_ROOT=os.path.join(os.environ['DYNAMO_HOME'], 'ext', 'SDKs')
+# The goal is to put the sdk versions in sdk.py
+SDK_ROOT=sdk.SDK_ROOT
+
 ANDROID_ROOT=SDK_ROOT
 ANDROID_BUILD_TOOLS_VERSION = '30.0.3'
-ANDROID_NDK_VERSION='20'
 ANDROID_NDK_API_VERSION='16' # Android 4.1
-ANDROID_NDK_ROOT=os.path.join(os.environ['DYNAMO_HOME'], 'ext', 'SDKs','android-ndk-r%s' % ANDROID_NDK_VERSION)
+ANDROID_NDK_ROOT=os.path.join(SDK_ROOT,'android-ndk-r%s' % sdk.ANDROID_NDK_VERSION)
```
Comment (JCash) [inline build_tools/waf_dynamo.py:74]: So far, I've only moved the Darwin platform to the sdk.py

File: build_tools/waf_dynamo.py [hunk 7]
```diff
@@ -64,33 +65,22 @@ def transform_runnable_path(platform, path):
...
-# (MinimumOSVersion and perhaps DTPlatformVersion)
-MIN_IOS_SDK_VERSION="8.0"
-
-OSX_SDK_VERSION="11.3"
-MIN_OSX_SDK_VERSION="10.7"
-
-XCODE_VERSION="12.5"
-XCODE_CLANG_VERSION="12.0.5"
-SWIFT_VERSION="5.0"
+CLANG_VERSION='clang-13.0.0'
```
Comment (JCash) [inline build_tools/waf_dynamo.py:81]: Might be moved later.

File: build_tools/waf_dynamo.py [hunk 8]
```diff
@@ -342,6 +331,9 @@ def default_flags(self):
         else:
             self.env.append_value(f, ['-DDM_PLATFORM_32BIT'])
 
+    if not hasattr(self, 'sdkinfo'):
```
Comment (JCash) [inline build_tools/waf_dynamo.py:334]: An optimization to not look for local tools too many times (took an extra 30s in total!)

File: engine/dlib/src/dlib/jsmn/jsmn.c [hunk 9]
```diff
@@ -65,7 +69,7 @@ static jsmnerr_t jsmn_parse_primitive(jsmn_parser *parser, const char *js,
 found:
 	if (tokens == NULL) {
 		parser->pos--;
-		return 0;
+		return JSMN_OK;
```
Comment (JCash) [inline engine/dlib/src/dlib/jsmn/jsmn.c:72]: Moving from clang 8 to 13 made the code warnings into errors.

File: engine/dlib/src/test/test_memprofile.cpp [hunk 10]
```diff
@@ -103,6 +103,7 @@ TEST(dmMemProfile, TestRealloc)
     dmMemProfile::Stats stats1, stats2, stats3;
     dmMemProfile::GetStats(&stats1);
     void* p = malloc(1024);
+    g_dont_optimize = p;
```
Comment (JCash) [inline engine/dlib/src/test/test_memprofile.cpp:106]: Clang 9 didn't optimize this away, but clang 13 did.

**Review by britzl — APPROVED** [review]

**Pull request merged**